File: app/src/main/java/es/etologic/mahjongscoring2/domain/use_cases/GetAllGamesUseCase.java
package es.etologic.mahjongscoring2.domain.use_cases;

import java.util.List;

import javax.inject.Inject;

import es.etologic.mahjongscoring2.data.repositories.GamesRepository;
import es.etologic.mahjongscoring2.domain.entities.Game;
import io.reactivex.Single;

public class GetAllGamesUseCase {

    private GamesRepository gamesRepository;

    @Inject
    public GetAllGamesUseCase(GamesRepository gamesRepository) { this.gamesRepository = gamesRepository; }

    public Single<List<Game>> getAll() { return gamesRepository.getAll(); }
}

File: app/src/main/java/com/etologic/mahjongscoring2/business/use_cases/GetAllGamesUseCase.kt
/*
 *     Copyright © 2023  Ernesto Vega de la Iglesia Soria
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.etologic.mahjongscoring2.business.use_cases

import com.etologic.mahjongscoring2.business.model.entities.UIGame
import com.etologic.mahjongscoring2.data_source.repositories.GamesRepository
import com.etologic.mahjongscoring2.data_source.repositories.RoundsRepository
import io.reactivex.Single
import javax.inject.Inject

class GetAllGamesUseCase @Inject constructor(
    private val gamesRepository: GamesRepository,
    private val roundsRepository: RoundsRepository,
) {
    operator fun invoke(): Single<List<UIGame>> =
        Single.zip(
            gamesRepository.getAll(),
            roundsRepository.getAll(),
        ) { games, rounds ->
            games.map { game ->
                val gameRounds = rounds.filter { round -> round.gameId == game.gameId }
                UIGame(game, gameRounds)
            }
        }
}
